pkg/ccl/backupccl/backup_planning.go

```diff
@@ -1077,26 +1081,94 @@ func getScheduledBackupExecutionArgsFromSchedule(
 	return sj, args, nil
 }
 
+// writeBackupManifestCheckpoint writes a new BACKUP-CHECKPOINT MANIFEST
+// and CHECKSUM file. If it is a pure v22.1 cluster or later, it will write
+// a timestamped BACKUP-CHECKPOINT to the /progress directory.
+// If it is a mixed cluster version, it will write a non timestamped BACKUP-CHECKPOINT
+// to the base directory in order to not break backup jobs that resume
+// on a v21.2 node.
 func writeBackupManifestCheckpoint(
 	ctx context.Context,
-	jobID jobspb.JobID,
-	backupDetails jobspb.BackupDetails,
-	backupManifest *BackupManifest,
+	storageURI string,
+	encryption *jobspb.BackupEncryptionOptions,
+	desc *BackupManifest,
 	execCfg *sql.ExecutorConfig,
 	user security.SQLUsername,
 ) error {
-	defaultStore, err := execCfg.DistSQLSrv.ExternalStorageFromURI(ctx, backupDetails.URI, user)
+	defaultStore, err := execCfg.DistSQLSrv.ExternalStorageFromURI(ctx, storageURI, user)
 	if err != nil {
 		return err
 	}
 	defer defaultStore.Close()
 
-	if err := writeBackupManifest(
-		ctx, execCfg.Settings, defaultStore, tempCheckpointFileNameForJob(jobID),
-		backupDetails.EncryptionOptions, backupManifest,
-	); err != nil {
-		return errors.Wrapf(err, "writing checkpoint file")
+	sort.Sort(BackupFileDescriptors(desc.Files))
+
+	descBuf, err := protoutil.Marshal(desc)
+	if err != nil {
+		return err
+	}
+
+	descBuf, err = compressData(descBuf)
+	if err != nil {
+		return errors.Wrap(err, "compressing backup manifest")
 	}
+
+	if encryption != nil {
+		encryptionKey, err := getEncryptionKey(ctx, encryption, execCfg.Settings, defaultStore.ExternalIOConf())
+		if err != nil {
+			return err
+		}
+		descBuf, err = storageccl.EncryptFile(descBuf, encryptionKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	// We timestamp the checkpoint files in order to enforce write once
+	// backups. By timestamping with monotonically increasing timestamps
+	// we avoid ever overwriting a checkpoint file.
+	filename := newTimestampedCheckpointFileName()
+
+	// HTTP storage does not support listing, and so the very first time
+	// we write a checkpoint we just write a non-versioned checkpoint file
+	// that we can always find without listing.
+	if defaultStore.Conf().Provider == roachpb.ExternalStorageProvider_http {
+		if r, err := defaultStore.ReadFile(ctx, backupProgressDirectory+"/"+backupManifestCheckpointName); err != nil {
+			filename = backupManifestCheckpointName
```

Comment:
I don't like that we are doing this unconditionally for all errors. Can you add a TODO to catch specific file not found or directory not found errors here?

Also a comment saying, since we did not find the checkpoint file this is the first time we are going to write a checkpoint, so write it with the well known filename.